### PR TITLE
Rework to use on Linux

### PR DESCRIPTION
**Pull request by AlwinEsch**
This adds code for processing OpenGL.

In addition, the settings are extended to include your own files from Shadertoy.

Furthermore fanart and screenshots are added, which the user sees what it brings.

In the shaders a lot had to be changed, so I'm not sure if they still work under Windows.

But noticed that in the addon no needed textures of the Shadertoy shaders are available and probably never really worked properly.

If someone has DirextX knowledges he might be able to attack it.

### CODE REVIEW

**Comment by afedchin** [comment]
did you test this on windows? because shaders were adapted and changes may break them

**Comment by AlwinEsch** [comment]
No, unfortunately not. I do not have a windows build system here. How did it behave before because Windows did not load the required textures?

In the shaders "static" was everywhere, this was not supported by OpenGL, is this "static" necessary under DirectX?

**Comment by AlwinEsch** [comment]
On the Windows C++ is mostly nothing changed, only taken from ShaderToy the code of the glsl files thats why the changes looks there bigger.

Some code differences were in the glsl, but it was assumed that the code was updated on the website.

**Review by afedchin — CHANGES_REQUESTED** [review]
> How did it behave before because Windows did not load the required textures?

it doesn't use textures

> In the shaders "static" was everywhere, this was not supported by OpenGL, is this "static" necessary under DirectX?

iirc static is required for `const`.

I had a quick look into shaders changes, there are a lot of breaking changes:
1. HLSL can't multiply matrixes with operator `*` there is operator `muv` for this. removing `mul` makes shaders uncompilable.
2. `vec3(1.0)` (and similar) doesn't work on windows. only `vec3(1.0, 1.0, 1.0)` will (I've replaced each similar occerences manually)
3. HLSL on FL9.x has limitation on amount of shader operations and loops with 100 iterations may not work, because compiler unfolds loops.

File: screensaver.shadertoy/resources/firefast.frag.glsl.orig [hunk 1]
```diff
@@ -1,27 +0,0 @@
-// Taken from https://www.shadertoy.com/view/4dlGR2
```
[the remark this answers is not in the document]
Comment (AlwinEsch) [reply screensaver.shadertoy/resources/firefast.frag.glsl.orig]: With *.orig says to me that something in the past was create by a "patch ..." call. Also was firefast not used.

[the remark this answers is not in the document]
Comment (afedchin) [reply screensaver.shadertoy/resources/firefast.frag.glsl.orig]: you are right, no need keep them. orig means original shader code (from the site)

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      compiler: gcc
+    - os: linux
+      dist: xenial
+      sudo: required
+      compiler: clang
+    - os: osx
+      osx_image: xcode9
+
+#
+# Some of the OS X images don't have cmake, contrary to what people
```
Comment (Rechi) [inline .travis.yml]: That's not true anymore.

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+# on the Internet say
+#
+before_install:
+  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then which cmake || brew update        ; fi
```
Comment (Rechi) [inline .travis.yml]:
```suggestion
```

File: .travis.yml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then which cmake || brew install cmake ; fi
```
Comment (Rechi) [inline .travis.yml]:
```suggestion
```

File: .travis.yml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew update && brew upgrade cmake || true; fi
```
Comment (Rechi) [inline .travis.yml]:
```suggestion
```

File: CMakeLists.txt [hunk 6]
```diff
@@ -1,22 +1,58 @@
...
+    set(includes ${OPENGL_INCLUDE_DIR})
+    add_definitions(${OPENGL_DEFINITIONS})
+  elseif(APP_RENDER_SYSTEM STREQUAL "gles")
+    find_package(OpenGLES REQUIRED)
+    set(DEPLIBS ${OPENGLES_LIBRARIES})
+    set(includes ${OPENGLES_INCLUDE_DIR})
+    add_definitions(${OPENGLES_DEFINITIONS})
+  endif()
+
+  if(NOT OPENGL_FOUND AND NOT OPENGLES_FOUND)
```
Comment (Rechi) [inline CMakeLists.txt]: Make this the else branch in the above if.

**Comment by AlwinEsch** [comment]
A small question.
After the "Angle" is in Kodi, do we switch this to primary as GL or leaving with the to DX translated shader code? If yes then I cleanup this after them is in Kodi.

**Comment by Rechi** [comment]
@AlwinEsch this needs a rebase

**Comment by AlwinEsch** [comment]
Is also in my workflow :grin:

**Pull request closed without merging**